### DISCUSSION TITLE
read and use opus R128 loudness data

### DISCUSSION THREAD

**Discussion by Almoped** [audacious-media-player/audacious#101] [Ideas]
Apparently opus uses EBU R 128, and not ReplayGain, tags for loudness equalization. They're stored in Q7.8 numbers and R128 is set 5 dB lower than ReplayGain 2.0. The attached patch reads the supported tags (RFC7845) and makes loudgained opus content play at the corrected levels while using Audacious'  gain adjustment implementation.

[opus loudness patch.txt](https://github.com/audacious-media-player/audacious/files/13933081/opus.loudness.patch.txt)

**Comment by radioactiveman** [comment]
`opusfile` offers API methods for replay gain. Could you please make use of them and submit a pull request?
Also attach a small sample file to test this. Thanks.

- opus_tags_get_track_gain()
- opus_tags_get_album_gain()

https://opus-codec.org/docs/opusfile_api-0.12/group__header__info.html

**Reply by Almoped** [reply]
opus_tags_get_track_gain() and related functions was the first thing I tried, but I could not make them work.
Yet opus_tags_query(tags, "R128_TRACK_GAIN", 0) seems to work just fine.

I've only got opus files tagged from loudgain, but they seem to work correctly with Foobar2000 as well.
Getting some sample files which are up to spec is probably the first step in getting the build-in opusfile functions working.

They do mention the functions should only be used to display the gain data, and have the decoder apply the gain adjustment internally, which is unlike how replay gain is applied for other formats in Audacious.

**Reply by radioactiveman** [reply]
I have pushed a fix based on your patch as https://github.com/audacious-media-player/audacious-plugins/commit/78f33f7316de481d6ce34ed3614076fda700aa1c, thanks for the groundwork.
